[MIR] Replace bespoke DIExpression parser

slinder1 opened the pull request:
Resolve FIXME by using the LLParser implementation of parseDIExpression from the MIParser.

slinder1 committed:
[MIR] Replace bespoke DIExpression parser

Resolve FIXME by using the LLParser implementation of parseDIExpression
from the MIParser.

slinder1 committed:
[NFC] Add DIExpression MIParser tests

Prepare for patch to replace MIParser implementation of DIExpression
parser with the LLParser version.

OCHyams commented on llvm/unittests/AsmParser/AsmParserTest.cpp:
```diff
@@ -411,4 +412,49 @@ TEST(AsmParserTest, InvalidDataLayoutStringCallback) {
+  ASSERT_TRUE(Expr);
+  ASSERT_EQ(Expr->getNumElements(), 3);
+
+  Expr = parseDIExpressionAtBeginning("(DW_OP_LLVM_fragment, 0, 1)", Read, Error, M, &Mapping);
+  ASSERT_TRUE(Expr);
+  ASSERT_EQ(Expr->getNumElements(), 3);
```
Why isn't this one ASSERT_FALSE?

slinder1 replied:
So this and the change of an `assert` in the parser to an `if` are to bridge the gap between the respective lexers for the MIParser (which consumes the `!DIExpression` token already by the time it would call `parseDIExpressionAtBeginning`) and LLParser (which does not consume the `!DIExpression` token by the time it would call `parseDIExpression`).

The alternative is to make both agree, so e.g. MIParser could just `peek` at the token or LLParser could `lex` it. I don't have a strong opinion on which approach is best.

OCHyams replied:
I also don't have a strong opinion on whether too peek or lex in one or the other, but I do think that `parseDIExpressionAtBeginning` should only accept either the whole thing or not (unless there's precedent for his pattern elsewhere, or it involves a lot of hoop-jumping to support).

OCHyams reviewed (commented):
Hi @scott-linder,

Overall SGTM but I've got a couple of nit/questions inline.

It's been a while since I've dug around in the parser so might be worth another pair of eyes glancing at this too.

slinder1 commented on llvm/test/CodeGen/MIR/Generic/diexpression-err.mir:
```diff
@@ -0,0 +1,152 @@
+# RUN: not llc -run-pass none -o - %s 2>&1 | FileCheck %s
+# Note: generated via:
```
Yes, I wanted to ensure the diagnostic retains essentially the same character before and after the change.

The diff between the first and second commit in the PR shows that the error message changes (with both versions being equivalent) but the line+column in the diagnostic is identical.

I didn't trim the test as I couldn't find a great reference for what extent to do so (e.g. https://groups.google.com/g/llvm-dev/c/6heEIwca_t0 doesn't seem to have a resolution wrt. the metadata itself), and testing "real world" input IR seemed the least likely to mistakenly verify something that isn't true in practice. I seem to run into this every time I write a test, and still haven't come up with a great approach.

OCHyams replied:
For this one, which looks like it really only needs a single DBG_VALUE with a broken expression, I'd probably just write it by hand. A method that avoids hand-writing the needed metadata might look like this: 

1. Compile an empty function to IR
2. Run `opt -passes=debugify` on it
3. Remove !dbg attachments from instructions, remove all but 1 of the dbg.values if there are multiple
4. llc -stop-after=...

And I think we could probably combine both tests into one file using a `sed` command to replace the bad DW_OP_... with a correctly spelled one. There should be `sed` examples in some of the llvm/test/DebugInfo tests iirc (you can test different outputs in the same test using the FileCheck `--check-prefixes` argument).

slinder1 committed:
Fold redundant `if` into return statement.

Co-authored-by: Orlando Cazalet-Hyams <orlandoch.och@gmail.com>

slinder1 committed:
Exclusively support parsing the "body" of a DIExpression

Simplify and extend tests, and limit "AtBeginning" parsing
to strings which do not include the leading "!DIExpression".

Change-Id: I62764d7d55667cd9036e0f3c484a1ab1363ef48e

slinder1 committed:
run clang-format

Change-Id: I0095ed73eed66d16b6ab35407bc655e76f06fead

slinder1 committed:
Remove redundant diexpression-err.mir test

Change-Id: I274f51393f33920f1246d32d84e2e3ff36cb66d2

slinder1 committed:
Improve tests and clang-format

OCHyams commented on llvm/lib/CodeGen/MIRParser/MIParser.cpp:
```diff
@@ -2302,48 +2302,14 @@ bool MIParser::parseMDNode(MDNode *&Node) {
-
-  Expr = DIExpression::get(MF.getFunction().getContext(), Elements);
+  unsigned Read;
+  Expr = llvm::parseDIExpressionBodyAtBeginning(
```
Does this need to be renamed if it doesn't parse a DIExpression from the beginning but from the parens onwards?

slinder1 replied:
The `AtBeginning` relates to the fact that the buffer can be longer than the body of the `DIExpression`, i.e. there can be trailing characters in the buffer. I intended to add a test for that property but missed it, so I'll add one and update the test.

OCHyams reviewed (commented):
I think this LGTM now, ty. It still would be good if someone else could glance at it too, cc @SLTozer who has poked around in the parser more recently.

slinder1 committed:
Add test for AtBeginning behavior

SLTozer reviewed (approved):
I see no further issues with this one!
